### PR TITLE
Publish Developer Guide

### LINKED ISSUE #235: [DOC] Improve doxygen rendering, grouping, etc.
We should follow the lead of libcudf and improve our doxygen docs

 - [x] Add a main_page.md 
 - [x] Add API groups and a tab for them
 - [x] tweak various doxygen settings to improve rendering / usability

### PR DESCRIPTION
**Pull request by harrism**
## Description

This PR mimics https://github.com/rapidsai/cudf/pull/11475 to add libcuspatial's C++ developer guide to the rendered Doxygen HTML docs. See that PR for detailed discussion.

Contributes to #598 . Closes #235

Note that this PR depends on #672 and #667 so the changed files will include changes from those PRs until they are merged.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cuspatial/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: ci/release/update-version.sh [hunk 1]
```diff
@@ -45,3 +45,7 @@ sed_runner 's/'"branch-.*\/RAPIDS.cmake"'/'"branch-${NEXT_SHORT_TAG}\/RAPIDS.cma
 for FILE in conda/environments/*.yml; do
   sed_runner "s/cudf=${CURRENT_SHORT_TAG}/cudf=${NEXT_SHORT_TAG}/g" ${FILE};
 done
+
+# Doxyfile update
+sed_runner "s|\(TAGFILES.*librmm/\).*|\1${NEXT_SHORT_TAG}|" cpp/doxygen/Doxyfile
+sed_runner "s|\(TAGFILES.*libcudf/\).*|\1${NEXT_SHORT_TAG}|" cpp/doxygen/Doxyfile
```
Comment (ajschmidt8) [inline ci/release/update-version.sh:51]:
```suggestion
sed_runner "/TAGFILES/ s|[0-9]\+.[0-9]\+|${NEXT_SHORT_TAG}|g" cpp/doxygen/Doxyfile
```

The original changes here didn't work for me when I tested them locally. This suggested change seems to do the trick. You can verify this locally by running `bash ci/release/update-version.sh 22.10.00` and verifying the results.

**Review by ajschmidt8 — CHANGES_REQUESTED** [review]

File: cpp/doxygen/developer_guide/BENCHMARKING.md [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+Both NVBench and Google Benchmark provide many options for specifying ranges of parameters to
+benchmark, as well as to control the time unit reported, among other options. Refer to existing
+benchmarks in `cpp/benchmarks` to understand the options.
+
+## Directory and File Naming
+
+The naming of unit benchmark directories and source files should be consistent with the feature
+being benchmarked. For example, the benchmarks for APIs in `point_in_polygon.hpp` should live in
+`cpp/benchmarks/point_in_polygon.cu`. Each feature (or set of related features) should have its own
+benchmark source file named `<feature>.cu/cpp`.
```
Comment (thomcom) [inline cpp/doxygen/developer_guide/BENCHMARKING.md:20]:
```suggestion
benchmark source file named `<feature>{.cu,cpp}`. 
```

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 3]
```diff
@@ -0,0 +1,530 @@
...
+sentence description. Doxygen does not provide much space to show this text in the output pages.
+Always follow the \@brief line with a blank comment line.
+
+The longer description is the rest of the comment text that is not tagged with any doxygen command.
+
+  /**
+   * @brief Short description.
+   *
+   * Long description.
+   *
```
Comment (thomcom) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:224]:
```suggestion
   */
```

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 4]
```diff
@@ -0,0 +1,530 @@
...
+\@addtogroup with `@{ ... @}` pair includes a namespace declaration. So include the \@addtogroup
+and `@{ ... @}` between the namespace declaration braces as shown in the example above.
+
+Summary of groups tags
+| Tag/Command | Where to use |
+| ----------- | ------------ |
+| \@defgroup | For use only in [doxygen_groups.h](../include/doxygen_groups.h) and should include the group's title. |
+| \@ingroup | Use inside individual doxygen block comments for declaration statements in a header file. |
+| \@addtogroup | Use instead of \@ingroup for multiple declarations in the same file within a namespace declaration. Do not specify a group title. |
+| `@{ ... @}` |  Use only with \@addtogroup. |
```
Comment (thomcom) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:507]: I think a list of the groups in `doxygen_groups.h` and their purpose might be appropriate here. Or a link to it telling the user to go read them.

**Review by thomcom — APPROVED** [review]

**Review by vyasr — APPROVED** [review]
I just focused on reviewing the changes to doxygen files and the associated scripts. Everything there looks fine, and the link added to the `DEVELOPER_GUIDE.md` should be sufficient for the built documentation to render appropriately. Probably worth trying to build locally once (in case you haven't yet) just to be sure, but I think this looks good.

File: cpp/doxygen/developer_guide/DOCUMENTATION.md (hunk 4 above)
Comment (harrism) [reply cpp/doxygen/developer_guide/DOCUMENTATION.md:507]: Done.

File: ci/release/update-version.sh [hunk 5]
```diff
@@ -45,3 +45,7 @@ sed_runner 's/'"branch-.*\/RAPIDS.cmake"'/'"branch-${NEXT_SHORT_TAG}\/RAPIDS.cma
 for FILE in conda/environments/*.yml; do
   sed_runner "s/cudf=${CURRENT_SHORT_TAG}/cudf=${NEXT_SHORT_TAG}/g" ${FILE};
 done
+
+# Doxyfile update
+sed_runner "/PROJECT_NUMBER/ s|[0-9]\+.[0-9]\+|${NEXT_SHORT_TAG}|g" cpp/doxygen/Doxyfile
```
Comment (ajschmidt8) [inline ci/release/update-version.sh:50]:
```suggestion
sed_runner "/PROJECT_NUMBER[ ]*=/ s|=.*|= ${NEXT_FULL_TAG}|g" cpp/doxygen/Doxyfile
```
One more update needed here. You can test this with various version numbers to ensure that everything works as intended. For example, run the following and check the `diff` each time.

```sh
bash ci/release/update-version.sh 22.12.00
bash ci/release/update-version.sh 22.12.02
```

**Review by ajschmidt8 — APPROVED** [review]

**Comment by harrism** [comment]
rerun tests

**Comment by harrism** [comment]
Depends on #667

**Comment by harrism** [comment]
Rerun tests

File: cpp/doxygen/Doxyfile [hunk 6]
```diff
@@ -726,7 +726,7 @@ FILE_VERSION_FILTER    =
 # DoxygenLayout.xml, doxygen will parse it automatically even if the LAYOUT_FILE
```
Comment (isVoid) [inline cpp/doxygen/Doxyfile:726]: `DoxygenLayout.xml` is default?

File: cpp/doxygen/developer_guide/DOCUMENTATION.md [hunk 7]
```diff
@@ -0,0 +1,537 @@
+# libcuspatial C++ Documentation Guide
```
Comment (isVoid) [inline cpp/doxygen/developer_guide/DOCUMENTATION.md:1]: Curious why this file is not labeled as `rename` but as addition. Assuming content is unchanged besides @thomcom 's [comment](https://github.com/rapidsai/cuspatial/pull/673/files#r975802135)

**Review by isVoid — APPROVED** [review]
Sorry late to the game, all looks good. Using `modify_fences.sh` to allow proper display in both github and published doxygen certainly is an interesting technique.

**Comment by isVoid** [comment]
@gpucibot merge

**Pull request merged**